net/tcp/tcp_sendfile.c

```diff
@@ -264,15 +275,42 @@ static uint16_t sendfile_eventhandler(FAR struct net_driver_s *dev,
           goto end_wait;
         }
 
-      /* No. Fall through to send more data if necessary */
+      /* Fast Retransmit (RFC 5681): an acknowledgment is considered a
+       * "duplicate" when (a) the receiver of the ACK has outstanding data,
+       * (b) the incoming acknowledgment carries no data, (c) the SYN and
+       * FIN bits are both off, (d) the acknowledgment number is equal to
+       * the greatest acknowledgment received on the given connection
+       * and (e) the advertised window in the incoming acknowledgment equals
+       * the advertised window in the last incoming acknowledgment.
+       */
+
+      if (pstate->snd_acked < pstate->snd_sent &&
+          (flags & TCP_NEWDATA) == 0 &&
+          (tcp->flags & (TCP_SYN | TCP_FIN)) == 0 &&
+          ackno == pstate->snd_prev_ack &&
+          conn->snd_wnd == pstate->snd_prev_wnd)
+        {
+          if (++pstate->snd_dup_acks >=
```

Comment:
No, we should not care. According to RFC 5681 the maximum number of duplicated ACKs is 3. We have CONFIG_NET_TCP_FAST_RETRANSMIT_WATERMARK parameter in NuttX (I do not know why it is not a constant), it is 3 by default. Anyway, CONFIG_NET_TCP_FAST_RETRANSMIT_WATERMARK is not reasonable to set very high (no profit). I would remove CONFIG_NET_TCP_FAST_RETRANSMIT_WATERMARK option at all because the existence of the option contradicts to RFC 5681.

Reply:
@pkarashchenko in my previous message I meant rollover if CONFIG_NET_TCP_FAST_RETRANSMIT_WATERMARK value is set larger than pstate->snd_dup_acks variable can store. E.g. if int is 16 bits (for some arch) and CONFIG_NET_TCP_FAST_RETRANSMIT_WATERMARK is set to e.g. 32769 (that is not reasonable at all), then pstate->snd_dup_acks >= CONFIG_NET_TCP_FAST_RETRANSMIT_WATERMARK condition will never trigger, and pstate->snd_dup_acks will rollover from 32767 to -32768.
Possibly you meant another case: CONFIG_NET_TCP_FAST_RETRANSMIT_WATERMARK is set normally to 3 (default). Once pstate->snd_dup_acks >= CONFIG_NET_TCP_FAST_RETRANSMIT_WATERMARK condition has triggered, and ackno == pstate->snd_prev_ack condition is true and not change for the next 2^15 iterations (that is normally impossible on practice), then pstate->snd_dup_acks will rollover from 32767 to -32768.
For the second case I have just added "pstate->snd_dup_acks = 0;" line if pstate->snd_dup_acks >= CONFIG_NET_TCP_FAST_RETRANSMIT_WATERMARK condition has triggered to make sure the counter does not count anymore.

Reply:
Had a brief look from mobile. Probably will have additional questions. Give me time to review on a full screen till Monday

Reply:
> No, we should not care. According to RFC 5681 the detected number of duplicated ACKs is exactly 3. We have CONFIG_NET_TCP_FAST_RETRANSMIT_WATERMARK parameter in NuttX (I do not know why it is not a constant), it is 3 by default. Anyway, CONFIG_NET_TCP_FAST_RETRANSMIT_WATERMARK is not reasonable to set very high (no profit but only harmful). I would remove CONFIG_NET_TCP_FAST_RETRANSMIT_WATERMARK option at all because the existence of the option contradicts to RFC 5681.

@pkarashchenko FYI concerning the first case I am preparing a new PR to transform CONFIG_NET_TCP_FAST_RETRANSMIT_WATERMARK integer option to a boolean one that will just enable / disable the Fast Retransmit algorithm. And I will add a constant equaled 3 as I wrote above. Thus let's think for PR #5311 that CONFIG_NET_TCP_FAST_RETRANSMIT_WATERMARK is always 3. I will create the new PR (CONFIG_NET_TCP_FAST_RETRANSMIT_WATERMARK option transformation) right after PR #5311 is merged to prevent collision.

Reply:
I think it is good to rework to
```
if (pstate->snd_dup_acks >=
    CONFIG_NET_TCP_FAST_RETRANSMIT_WATERMARK)
  {
    flags |= TCP_REXMIT;
  }
else
  {
    pstate->snd_dup_acks++;
  }
```
@a-lunev  what do you think?

Reply:
Maybe it is personally me, but I do not like `if (++pstate->snd_dup_acks >= CONFIG_NET_TCP_FAST_RETRANSMIT_WATERMARK)` that potentially can count endlessly with the each check. I read your explanation and understand that it probably will never happen, but such code reminds me about the bugs that I spend dozens of hours to find the root cause.

Reply:
Yes, the same is in tcp_send_unbuffered.c and tcp_send_buffered.c (and e.g. here in another project: https://github.com/dferbas/ethernut/blob/5dd3894c942ce0c627ea8cc2666b3d99bacb114b/nut/net/tcpsm.c#L437)
I understand your feeling. If we follow your suggestion, the code should be e.g. as follows (to compensate the pre-increment in the original version):
```
if (pstate->snd_dup_acks >=
    CONFIG_NET_TCP_FAST_RETRANSMIT_WATERMARK - 1)
  {
    flags |= TCP_REXMIT;
  }
else
  {
    pstate->snd_dup_acks++;
  }
```
However, in case of the triggered condition pstate->snd_dup_acks will freeze at 2 (instead of 3) that may be confusing at least during debug. Or we should pre-initialize pstate->snd_dup_acks to 1 instead of 0, that is also is going to be confusing.
Thus I'm more inclining towards the original version (++pstate->snd_dup_acks >=CONFIG_NET_TCP_FAST_RETRANSMIT_WATERMARK).

What do you think?

Reply:
Ok. We can keep existing version. Or add `pstate->snd_dup_acks = CONFIG_NET_TCP_FAST_RETRANSMIT_WATERMARK;` in case if retransmission condition is met. Just for safety reasons.

Reply:
> @pkarashchenko FYI concerning the first case I am preparing a new PR to transform CONFIG_NET_TCP_FAST_RETRANSMIT_WATERMARK integer option to a boolean one that will just enable / disable the Fast Retransmit algorithm. And I will add a constant equaled 3 as I wrote above. Thus let's think for PR #5311 that CONFIG_NET_TCP_FAST_RETRANSMIT_WATERMARK is always 3. I will create the new PR (CONFIG_NET_TCP_FAST_RETRANSMIT_WATERMARK option transformation) right after PR #5311 is merged to prevent collision.

As I wrote I have created that new PR (#5341) to transform CONFIG_NET_TCP_FAST_RETRANSMIT_WATERMARK option.